INT-3860: TCP - Fix Port Conflicts

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3860

Similar to INT-3794, if some other process is bound to all adapters

```
$ lsof -i tcp -nP | grep 9000
java       6604 grussell  380u  IPv6 0x407e749ec47fa337      0t0  TCP *:9000 (LISTEN)
```

The socket utils reports that the localhost is available for binding; however, the test then
binds to all adapters.
1. Where (easily) possible, change tests to use port=0 and let the OS choose.
2. For other tests, use high-numbered ports.
3. Change tests to explicitly bind to localhost (like the utility did).
4. Fix `tcpOutboundGatewayInsideChain` test to use the correct port.

garyrussell committed:
INT-3860: TCP - Fix Port Conflicts

JIRA: https://jira.spring.io/browse/INT-3860

Similar to INT-3794, if some other process is bound to all adapters

    $ lsof -i tcp -nP | grep 9000
    java       6604 grussell  380u  IPv6 0x407e749ec47fa337      0t0  TCP *:9000 (LISTEN)

The socket utils reports that the localhost is available for binding; however, the test then
binds to all adapters.

1. Where (easily) possible, change tests to use port=0 and let the OS choose.
2. For other tests, use high-numbered ports.
3. Change tests to explicitly bind to localhost (like the utility did).
4. Fix `tcpOutboundGatewayInsideChain` test to use the correct port.

artembilan commented on spring-integration-ip/src/test/java/org/springframework/integration/ip/tcp/TcpConfigOutboundGatewayTests-context.xml:
```diff
@@ -37,6 +38,9 @@
 
+	<ip:tcp-inbound-gateway id="gatewayCrLf2"
+		connection-factory="crLfServer2"
+		request-channel="toSA" />
```
Would mind explaining this trick?

I guess it is a part of:
> Fix tcpOutboundGatewayInsideChain test to use the correct port.

But isn't clear how it works.
Thanks.

garyrussell replied:
Perviously, the server used the same port as `crLfServerNio` (7400) and `crLfServer2` had no gateway so it was never started. The result was that this test actually talked to the other inbound gateway.

artembilan replied:
Got it! 	
Nasty bug, BTW...

I haven't pulled it locally yet for merging, but I wonder if that Travis failure is a bell for you.

garyrussell replied:
I tried re-running the travis build - to no avail - investigating...

garyrussell committed:
Polishing - Fix Travis

Bind SOLingerTests to localhost.

Since we're using port 0, we have to wait until listening before calling getPort().

garyrussell commented:
Pushed; I think this will fix it, but let's wait for travis.

artembilan commented:
All green. Merging...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/3118596427b2789bcbd134dd691c34af064d7574 after rebase and squash
